fix: Improve command tokenization to handle quotes and escapes

Khitoshi opened the pull request:
fix #834 

* Removed the simple tokenization using `strings.Fields(command)` and switched to the new **`tokenizeWithQuotes`** function.

  * Parses single (`'`) and double (`"`) quotes as well as back-slash escapes, allowing paths with spaces to be split safely.
  * Detects unterminated quotes or lone back-slashes and returns an `error`.

* Updated the call site in `getPromptAction()` to use `tokenizeWithQuotes`.

* Implemented entirely with the standard library—no additional dependencies.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Bug Fixes**
  * Improved command input handling to correctly process quoted substrings and escape characters, ensuring more accurate tokenization of user commands. Unmatched quotes or escape characters now result in clear error messages.
* **Tests**
  * Added extensive tests covering command tokenization with quotes, escapes, and error cases.
  * Enhanced prompt command tests to verify handling of directory names and shell commands with spaces and quotes.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

lazysegtree commented on src/internal/ui/prompt/tokenize.go:
```diff
@@ -7,6 +7,7 @@ import (
 	"os"
```
(1) High level comment . Please update unit tests to cover this behaviour
update tokenize_test.go
please include some tests in model_prompt_test.go for directory names with space and quotes .  (This test is end to end test for the entire prompt feature)

lazysegtree replied:
(2) Please add a unit test for shell substitution inside quotes

`cd "$HOME"` and  cd '$HOME' should work. I think we only have `cd $HOME` as a test.

Please add any other corner cases to make sure that there is no regression after further changes.

lazysegtree replied:
(3) 
- Make sure to have some testcase for tokenization of escaped quotes `\"`
- Make sure to consider invalide tokens like `abcd "sdf`, `"abcd'`

lazysegtree replied:
@coderabbitai you can also suggest some edge cases . But please be very concise. Dont make too big of a message

Khitoshi replied:
@coderabbitai
I fixed it. Check it out.

lazysegtree commented on src/internal/ui/prompt/tokenize.go:
```diff
@@ -110,3 +111,56 @@ func findEndingBracket(r []rune, openIdx int, openParan rune, closeParan rune) i
+
+	for _, r := range command {
+		switch {
+		case escaped:
```
So a string like "\n\t" will be tokenized into "nt". Is this acceptable here ?

If our purpose of allowing escaped sequences is just for specific characters like `\$ \' \" \\`, but not all, then we should explicitly implement that behaviour.

lazysegtree replied:
@coderabbitai what do you think about this ? 
Is there any alternative way of implementing the escaped characters. And what would be the use case for escaped characters for our prompt model ?

lazysegtree commented on src/internal/ui/prompt/tokenize.go:
```diff
@@ -110,3 +111,56 @@ func findEndingBracket(r []rune, openIdx int, openParan rune, closeParan rune) i
 }
+
+// splits command into tokens while respecting quotes and escapes
+func tokenizeWithQuotes(command string) ([]string, error) {
```
@coderabbitai Please review if this tokenization implementation is solid, or if there are any edge cases missing to be handled ? Dont try to make it too perfect, as our use case is rather limited. Focus on correctness, not supporting more features

lazysegtree commented:
Please merge main branch in your branch. Recent fixes for windows unit tests have just went in today.

Khitoshi commented:
Thanks a lot for the review!
I've applied the suggested changes and updated the code accordingly!


--- 
src/internal/ui/prompt/tokenize.go
* fix: Fixed an issue where unnecessary characters were being escaped.

src/internal/ui/prompt/tokenize_test.go
* feat: Add tests for tokenization with quotes and special characters

src/internal/model_prompt_test.go
* feat: Add cases for handling directory names and shell commands with spaces and quotes

Khitoshi commented:
```
>./dev.sh
🚀 Starting superfile development workflow

==> Tidying Go modules...
✓ Go modules tidied
==> Formatting Go code...
✓ Code formatted
==> Running golangci-lint...
✓ Linting passed
==> Running unit tests...
?       github.com/yorukot/superfile    [no test files]
?       github.com/yorukot/superfile/src/cmd    [no test files]
?       github.com/yorukot/superfile/src/config [no test files]
?       github.com/yorukot/superfile/src/config/icon    [no test files]
ok      github.com/yorukot/superfile/src/internal       (cached)
ok      github.com/yorukot/superfile/src/internal/common        (cached)
?       github.com/yorukot/superfile/src/internal/ui    [no test files]
ok      github.com/yorukot/superfile/src/internal/ui/prompt     (cached)
ok      github.com/yorukot/superfile/src/internal/ui/rendering  (cached)
ok      github.com/yorukot/superfile/src/internal/ui/sidebar    (cached)
ok      github.com/yorukot/superfile/src/internal/utils (cached)
?       github.com/yorukot/superfile/src/pkg/file_preview       [no test files]
?       github.com/yorukot/superfile/src/pkg/string_function    [no test files]
✓ Unit tests passed
==> Building spf binary...
✓ Build completed successfully

🎉 All steps completed successfully!
Binary location: ./bin/spf
Binary size: 23M
```

lazysegtree commented:
Looks like some test are failing specifically in windows. Windows is always a pain, with their strange behaviours. 

Try to fix it, I can try to help you out with the fix, but on weekends only. I am a bit busy this week.

lazysegtree reviewed (approved):
Thanks for the contribution and quickly following up with the required changes.

lazysegtree commented on src/internal/model_prompt_test.go:
```diff
@@ -5,6 +5,7 @@ import (
 	"os"
 	"path/filepath"
 	"runtime"
+	"strings"
```
@coderabbit

Please come up with some of the bad things in this PR

lazysegtree commented on src/internal/ui/prompt/tokenize.go:
```diff
@@ -110,3 +111,65 @@ func findEndingBracket(r []rune, openIdx int, openParan rune, closeParan rune) i
+		case escaped:
+			// Only allow escaping of specific characters that have special meaning
+			switch r {
+			case '"', '\'', '\\', ' ':
```
It is bad that `' '` is included in this case statement, because `\ ` is not a valid escape sequence. 

@coderabbitai Is that true ?
